File: subagentic_ai_py/subagentic_ai/front_line_agent.py
"""Front Line Agent (FLA) - orchestration layer."""

from datetime import datetime, timedelta
from typing import Dict, List, Optional
from uuid import uuid4

from .sub_agent import ISubAgent, SubAgent
from .communication_authorization import CommunicationAuthorization
from .trust_profile import TrustProfile


class FrontLineAgent:
    """
    Front Line Agent (FLA) - orchestrates SubAgents, enforces isolation, and manages communication.
    Acts as chief of staff without holding domain knowledge.
    """
    
    def __init__(self, agent_id: str, default_trust_profile: TrustProfile):
        """
        Create a new Front Line Agent.
        
        Args:
            agent_id: Unique identifier for this FLA
            default_trust_profile: Default trust profile for SubAgents
        """
        self._id = agent_id
        self._default_trust_profile = default_trust_profile
        self._active_sub_agents: Dict[str, ISubAgent] = {}
        self._communication_authorizations: List[CommunicationAuthorization] = []
        self._event_log: List[str] = []
        
        self._log_event(f"FLA created: {agent_id}")
    
    @property
    def id(self) -> str:
        return self._id
    
    def classify_domain(self, user_intent: str) -> str:
        """
        Interpret user intent and classify the domain.
        
        Args:
            user_intent: User's stated intent
            
        Returns:
            Classified domain
        """
        self._log_event(f"Classifying domain for intent: {user_intent}")
        
        # Simplified classification - in real implementation would use ML/NLP
        intent = user_intent.lower()
        
        if "data" in intent or "analyze" in intent:
            return "DataAnalysis"
        elif "code" in intent or "program" in intent:
            return "CodeGeneration"
        elif "security" in intent or "audit" in intent:
            return "SecurityAudit"
        elif "research" in intent or "find" in intent:
            return "Research"
        
        return "General"
    
    async def create_sub_agent(self, domain: str, 
                               trust_profile: Optional[TrustProfile] = None) -> ISubAgent:
        """
        Spin up a new SubAgent for a specific domain/mission.
        
        Args:
            domain: Domain or mission focus
            trust_profile: Optional custom trust profile
            
        Returns:
            Created SubAgent instance
        """
        agent_id = f"SA-{domain}-{uuid4().hex}"
        profile = trust_profile or self._default_trust_profile
        
        sub_agent = SubAgent(agent_id, domain, profile)
        self._active_sub_agents[agent_id] = sub_agent
        
        await sub_agent.start_mission()
        
        self._log_event(f"Created and started SubAgent: {agent_id} for domain: {domain}")
        
        return sub_agent
    
    async def reactivate_sub_agent(self, agent_id: str) -> Optional[ISubAgent]:
        """
        Reactivate an existing SubAgent (if it was previously created).
        
        Args:
            agent_id: SubAgent identifier
            
        Returns:
            SubAgent instance or None if not found
        """
        agent = self._active_sub_agents.get(agent_id)
        if agent:
            self._log_event(f"Reactivated SubAgent: {agent_id}")
        else:
            self._log_event(f"Failed to reactivate SubAgent: {agent_id} - not found")
        
        return agent
    
    def authorize_communication(self, from_agent_id: str, to_agent_id: str,
                               is_bidirectional: bool = False,
                               allowed_data_scopes: Optional[List[str]] = None,
                               ttl: Optional[timedelta] = None) -> None:
        """
        Authorize communication between two SubAgents.
        
        Args:
            from_agent_id: Source agent ID
            to_agent_id: Target agent ID
            is_bidirectional: Whether communication is two-way
            allowed_data_scopes: List of allowed data scopes
            ttl: Time to live for this authorization
        """
        scopes = tuple(allowed_data_scopes) if allowed_data_scopes else None
        authorization = CommunicationAuthorization.create(
            from_agent_id, to_agent_id, is_bidirectional, scopes, ttl
        )
        
        self._communication_authorizations.append(authorization)
        
        self._log_event(
            f"Authorized communication: {from_agent_id} -> {to_agent_id} "
            f"(bidirectional: {is_bidirectional})"
        )
    
    async def mediate_communication(self, from_agent_id: str, to_agent_id: str,
                                   information: str) -> bool:
        """
        Mediate communication between SubAgents (FLA-mediated communication).
        
        Args:
            from_agent_id: Source agent ID
            to_agent_id: Target agent ID
            information: Information to transmit
            
        Returns:
            True if communication succeeded
        """
        # Check if communication is authorized
        authorization = next(
            (auth for auth in self._communication_authorizations
             if auth.is_authorized(from_agent_id, to_agent_id)),
            None
        )
        
        if not authorization:
            self._log_event(
                f"Communication blocked: {from_agent_id} -> {to_agent_id} - not authorized"
            )
            return False
        
        sender = self._active_sub_agents.get(from_agent_id)
        if not sender:
            self._log_event(f"Communication failed: sender {from_agent_id} not found")
            return False
        
        receiver = self._active_sub_agents.get(to_agent_id)
        if not receiver:
            self._log_event(f"Communication failed: receiver {to_agent_id} not found")
            return False
        
        # Send through sender's outbound (security level check)
        sent_success = await sender.send_information(to_agent_id, information)
        if not sent_success:
            self._log_event("Communication failed: sender rejected outbound")
            return False
        
        # Receive at receiver's inbound (trust level check)
        receive_success = await receiver.receive_information(from_agent_id, information)
        
        self._log_event(
            f"Communication mediated: {from_agent_id} -> {to_agent_id} - success: {receive_success}"
        )
        
        return receive_success
    
    async def retire_sub_agent(self, agent_id: str) -> None:
        """
        Retire a SubAgent when its mission concludes.
        
        Args:
            agent_id: SubAgent identifier
        """
        agent = self._active_sub_agents.get(agent_id)
        if agent:
            await agent.retire()
            del self._active_sub_agents[agent_id]
            
            # Remove any communication authorizations involving this agent
            self._communication_authorizations = [
                auth for auth in self._communication_authorizations
                if auth.from_agent_id != agent_id and auth.to_agent_id != agent_id
            ]
            
            self._log_event(f"Retired SubAgent: {agent_id}")
    
    async def process_user_task(self, user_intent: str, task_details: str) -> str:
        """
        Process a user task by creating appropriate SubAgent(s).
        
        Args:
            user_intent: User's stated intent
            task_details: Details of the task
            
        Returns:
            Task result
        """
        self._log_event(f"Processing user task: {user_intent}")
        
        domain = self.classify_domain(user_intent)
        sub_agent = await self.create_sub_agent(domain)
        
        try:
            result = await sub_agent.process_task(task_details)
            await sub_agent.complete_mission()
            return result
        except Exception as e:
            self._log_event(f"Task processing failed: {str(e)}")
            raise
    
    def get_active_sub_agents(self) -> Dict[str, ISubAgent]:
        """Get all active SubAgents."""
        return self._active_sub_agents.copy()
    
    def get_event_log(self) -> List[str]:
        """Get the event log for auditing."""
        return self._event_log.copy()
    
    def _log_event(self, event_description: str) -> None:
        """Log an event."""
        log_entry = (
            f"[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]}] "
            f"[FLA-{self._id}] {event_description}"
        )
        self._event_log.append(log_entry)

File: SubAgenticAI.Core/FrontLineAgent.cs
namespace SubAgenticAI.Core;

/// <summary>
/// Front Line Agent (FLA) - orchestrates SubAgents, enforces isolation, and manages communication.
/// Acts as chief of staff without holding domain knowledge.
/// </summary>
public class FrontLineAgent
{
    private readonly Dictionary<string, ISubAgent> _activeSubAgents;
    private readonly List<CommunicationAuthorization> _communicationAuthorizations;
    private readonly List<string> _eventLog;
    private readonly TrustProfile _defaultTrustProfile;
    
    public string Id { get; }
    
    /// <summary>
    /// Creates a new Front Line Agent
    /// </summary>
    public FrontLineAgent(string id, TrustProfile defaultTrustProfile)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        _defaultTrustProfile = defaultTrustProfile ?? throw new ArgumentNullException(nameof(defaultTrustProfile));
        
        _activeSubAgents = new Dictionary<string, ISubAgent>();
        _communicationAuthorizations = new List<CommunicationAuthorization>();
        _eventLog = new List<string>();
        
        LogEvent($"FLA created: {Id}");
    }
    
    /// <summary>
    /// Interprets user intent and classifies the domain
    /// </summary>
    public string ClassifyDomain(string userIntent)
    {
        LogEvent($"Classifying domain for intent: {userIntent}");
        
        // This is a simplified classification - in real implementation would use ML/NLP
        var intent = userIntent.ToLowerInvariant();
        
        if (intent.Contains("data") || intent.Contains("analyze"))
            return "DataAnalysis";
        if (intent.Contains("code") || intent.Contains("program"))
            return "CodeGeneration";
        if (intent.Contains("security") || intent.Contains("audit"))
            return "SecurityAudit";
        if (intent.Contains("research") || intent.Contains("find"))
            return "Research";
        
        return "General";
    }
    
    /// <summary>
    /// Spins up a new SubAgent for a specific domain/mission
    /// </summary>
    public async Task<ISubAgent> CreateSubAgentAsync(string domain, TrustProfile? trustProfile = null)
    {
        var agentId = $"SA-{domain}-{Guid.NewGuid():N}";
        var profile = trustProfile ?? _defaultTrustProfile;
        
        var subAgent = new SubAgent(agentId, domain, profile);
        _activeSubAgents[agentId] = subAgent;
        
        await subAgent.StartMissionAsync();
        
        LogEvent($"Created and started SubAgent: {agentId} for domain: {domain}");
        
        return subAgent;
    }
    
    /// <summary>
    /// Reactivates an existing SubAgent (if it was previously created)
    /// </summary>
    public Task<ISubAgent?> ReactivateSubAgentAsync(string agentId)
    {
        if (_activeSubAgents.TryGetValue(agentId, out var agent))
        {
            LogEvent($"Reactivated SubAgent: {agentId}");
            return Task.FromResult<ISubAgent?>(agent);
        }
        
        LogEvent($"Failed to reactivate SubAgent: {agentId} - not found");
        return Task.FromResult<ISubAgent?>(null);
    }
    
    /// <summary>
    /// Authorizes communication between two SubAgents
    /// </summary>
    public void AuthorizeCommunication(
        string fromAgentId, 
        string toAgentId, 
        bool isBidirectional = false,
        List<string>? allowedDataScopes = null,
        TimeSpan? ttl = null)
    {
        var authorization = new CommunicationAuthorization(
            fromAgentId, 
            toAgentId, 
            isBidirectional, 
            allowedDataScopes, 
            ttl);
        
        _communicationAuthorizations.Add(authorization);
        
        LogEvent($"Authorized communication: {fromAgentId} -> {toAgentId} " +
                $"(bidirectional: {isBidirectional})");
    }
    
    /// <summary>
    /// Mediates communication between SubAgents (FLA-mediated communication)
    /// </summary>
    public async Task<bool> MediateCommunicationAsync(
        string fromAgentId, 
        string toAgentId, 
        string information)
    {
        // Check if communication is authorized
        var authorization = _communicationAuthorizations
            .FirstOrDefault(a => a.IsAuthorized(fromAgentId, toAgentId));
        
        if (authorization == null)
        {
            LogEvent($"Communication blocked: {fromAgentId} -> {toAgentId} - not authorized");
            return false;
        }
        
        if (!_activeSubAgents.TryGetValue(fromAgentId, out var sender))
        {
            LogEvent($"Communication failed: sender {fromAgentId} not found");
            return false;
        }
        
        if (!_activeSubAgents.TryGetValue(toAgentId, out var receiver))
        {
            LogEvent($"Communication failed: receiver {toAgentId} not found");
            return false;
        }
        
        // Send through sender's outbound (security level check)
        var sentSuccess = await sender.SendInformationAsync(toAgentId, information);
        if (!sentSuccess)
        {
            LogEvent($"Communication failed: sender rejected outbound");
            return false;
        }
        
        // Receive at receiver's inbound (trust level check)
        var receiveSuccess = await receiver.ReceiveInformationAsync(fromAgentId, information);
        
        LogEvent($"Communication mediated: {fromAgentId} -> {toAgentId} - success: {receiveSuccess}");
        
        return receiveSuccess;
    }
    
    /// <summary>
    /// Retires a SubAgent when its mission concludes
    /// </summary>
    public async Task RetireSubAgentAsync(string agentId)
    {
        if (_activeSubAgents.TryGetValue(agentId, out var agent))
        {
            await agent.RetireAsync();
            _activeSubAgents.Remove(agentId);
            
            // Remove any communication authorizations involving this agent
            _communicationAuthorizations.RemoveAll(a => 
                a.FromAgentId == agentId || a.ToAgentId == agentId);
            
            LogEvent($"Retired SubAgent: {agentId}");
        }
    }
    
    /// <summary>
    /// Processes a user task by creating appropriate SubAgent(s)
    /// </summary>
    public async Task<string> ProcessUserTaskAsync(string userIntent, string taskDetails)
    {
        LogEvent($"Processing user task: {userIntent}");
        
        var domain = ClassifyDomain(userIntent);
        var subAgent = await CreateSubAgentAsync(domain);
        
        try
        {
            var result = await subAgent.ProcessTaskAsync(taskDetails);
            await subAgent.CompleteMissionAsync();
            
            return result;
        }
        catch (Exception ex)
        {
            LogEvent($"Task processing failed: {ex.Message}");
            throw;
        }
    }
    
    /// <summary>
    /// Gets all active SubAgents
    /// </summary>
    public IReadOnlyDictionary<string, ISubAgent> GetActiveSubAgents() => 
        new Dictionary<string, ISubAgent>(_activeSubAgents);
    
    /// <summary>
    /// Gets the event log for auditing
    /// </summary>
    public IReadOnlyList<string> GetEventLog() => _eventLog.AsReadOnly();
    
    private void LogEvent(string eventDescription)
    {
        var logEntry = $"[{DateTime.UtcNow:yyyy-MM-dd HH:mm:ss.fff}] [FLA-{Id}] {eventDescription}";
        _eventLog.Add(logEntry);
    }
}
